### DISCUSSION TITLE
How could open the client-side socket ?

### DISCUSSION THREAD

**Discussion by dgasacas7** [jtv/libpqxx#899] [Q&A] [answered]
When trying to create a pqxx::connection object, the app crashes with an error message like this,

```
unhandled exception (type std::exception) in signal handler:
what: connection to server on socket "/var/run/postgresql/.s.PGSQL.5432" failed: No such file or directory
	Is the server running locally and accepting connections on that socket?
```

This error solves after starting locally a PostgreSQL server process/service... but the goal is to run both processes (client & server) in different machines...

Notice that the official client psql works fine in this distributed scenario.

I'm working with libpq (17.0) & libpqxx (7.9.2)

**Comment by tt4g** [comment]
Should you pass Connection URI string to `connection` constructor: https://www.postgresql.org/docs/17/libpq-connect.html#LIBPQ-CONNSTRING-URIS

**Reply by dgasacas7** [reply]
Obviously, I'm passing the string connection to the constructor, like

> [...] pqxx::connection connection (pg_conn_data.get_conn_string()); [...]

I moved from key/value format to URI's... and nothing changed. If PostgreSQL is running locally (client-side), the code runs fine... but it breaks when PostgreSQL is stopped.

**Reply by jtv** [reply]
Also... if the error says "unhandled exception," then start by handling the exception!

if you catch the exception and print its message, you will have a clearer idea of what's going on.  This is basic standard practice.

**Reply by dgasacas7** [reply]
Here goes my piece of code,
```
    try {
        connection = pqxx::connection (pg_conn_data.get_conn_string());
    } catch (const std::exception &e) {
        set_error_message (e.what());
    }
```
I tried to print the e.what() value without any success. Seems that code breaks before.

**Reply by tt4g** [reply]
It is assumed that the default constructor is invoked because it is not explicitly initialized where the `connection` variable is declared.

**Reply by dgasacas7** [reply]
`connection` variable is a class object and it is initialized in the constructor.

```
class PgSQL
{
public:

    PgSQL (const PgConnData _pg_conn_data)
    :
        pg_conn_data (_pg_conn_data)
    {
        try {
            connection = pqxx::connection (pg_conn_data.get_conn_string());
        } catch (const std::exception &e) {
            set_error_message (e.what());
        }
    }

    ~PgSQL();

    [...]

private:

    PgConnData pg_conn_data;
    pqxx::connection connection;
    std::string error_message;
};
```
What should I do to work it around ?
What do you suggest ?

**Reply by tt4g** [reply, marked as answer]
In C++, all member variables are initialized before the constructor body is evaluated.
If not explicitly initialized, the compiler calls the default constructor.

Example ([run online](https://wandbox.org/permlink/gUqcpPiJDGnui33z)):

```cpp
#include <iostream>
#include <cstdlib>

class Foo
{
public:
    Foo()
    {
        std::cout << "FOO initialized!" << std::endl;
    }

    explicit Foo(int)
    {
        std::cout << "FOO(int) initialized!" << std::endl;
    }
};

class Bar
{
public:
    Bar()
    { // Output: FOO initialized!
        foo = Foo(1); // Output: FOO(int) initialized!
    }
private:
    Foo foo;
};

int main()
{
    Bar bar;
}
```

To avoid this, initialize member variables before the constructor body: https://en.cppreference.com/w/cpp/language/constructor

Example ([run online](https://wandbox.org/permlink/Z9Sbsked68T2GwbE)):

```diff
 class Bar
 {
 public:
-    Bar()
-    { // Output: FOO initialized!
-        foo = Foo(1); // Output: FOO(int) initialized!
-    }
+    Bar() : foo(1) // // Output: FOO(int) initialized!
+    {
+    }
 private:
     Foo foo;
 };
```

If you really want to do initialization in the constructor body, use a wrapper object that can be initialized with empty data such as `std::optional` or `std::unique_ptr` as a member variable.
